Added SRP client implementation

kkasperczyk-no opened the pull request.

kkasperczyk-no commented:
Current implementation doesn't contain using SRP Client in CHIP stack yet. It is just an implementation of necessary methods and config options.

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +993,233 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+    otError error                   = OT_ERROR_NONE;
+    otSrpClientService * srpService = nullptr;
+
+    VerifyOrExit(aInstanceName, error = OT_ERROR_INVALID_ARGS);
```
Wouldn't it be confusing to see OT errors in case it's actually an issue in CHIP?

kkasperczyk-no replied:
Done

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +993,233 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+    // Convert server service config data to get from it server IP address and port number.
+    otSockAddr serverSocketAddress;
+    serverSocketAddress.mAddress =
+        GenericThreadStackManagerImpl_OpenThread::ConvertSrpServerRloc16ToIpAddress(serverServiceConfig.mServerConfig.mRloc16);
```
Is `GenericThreadStackManagerImpl_OpenThread::` prefix needed?

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +993,233 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+bool GenericThreadStackManagerImpl_OpenThread<ImplClass>::FindSrpServerService(otServiceConfig & aServiceConfig)
+{
+    bool serviceFound = false;
+    Impl()->LockThreadStack();
```
Shouldn't we rather guard the public methods like `AddSrpService` `RemoveSrpService`... ?

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +993,233 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+}
+
+template <class ImplClass>
+uint16_t GenericThreadStackManagerImpl_OpenThread<ImplClass>::ConvertSrpServerDataToPort(uint8_t * aServerData)
```
`static_assert(OT_SERVER_DATA_MAX_SIZE >= 2, "");` to convince a reader that it's safe to access the array elements?

LuDuda replied:
This adds some more confidence to the reader, but anyway I believe we should check the value of `mServerDataLength`, and see if it's equal/equal_or_greater than two, probably not here but when you execute this function.

kkasperczyk-no replied:
Both changes done

Damian-Nordic reviewed (commented):
I found one serious issue and have a few suggestions, but very good job so far!

kkasperczyk-no commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +993,233 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+    // Verify is there an empty place for new service.
+    VerifyOrExit(srpService, error = OT_ERROR_NO_BUFS);
+
+    memset(srpService, 0, sizeof(otSrpClientService));
```
And what if we call AddSrpService for the first time, so srpService will point to the first element of array, which has never been used before? I guess it might be filled with trash.

Damian-Nordic replied:
I think you should zero the array in the initialization function. Otherwise, you may never find a service which has never been used ;)

kkasperczyk-no commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +993,233 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
     return error;
 }
 
+#if CHIP_DEVICE_CONFIG_ENABLE_SRP_CLIENT && OPENTHREAD_API_VERSION >= 72
```
So according to this, we should remove #else from here: https://github.com/kkasperczyk-no/connectedhomeip/pull/21/files#diff-37cb9fb56dea71436c39deea1c9c7256f94d3afa909efb2d8ffebc5589e751eaR98 ?

Damian-Nordic replied:
Yes.

kkasperczyk-no commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +993,233 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+        break;
+    }
+    case OT_ERROR_PARSE:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Parsing operaton failed");
```
Do you mean logging just error code without explaining it?

Damian-Nordic replied:
It was just a thought that whatever chip::ErrorStr() would produce might be enough, but maybe you're right and it's better to explain the error in the context of SRP. Disregard this comment.

LuDuda commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +998,237 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+
+    // Replace device rloc16 bytes with the server rloc16
+    uint16_t addrSize                        = sizeof(serverIpAddress.mFields.m8);
+    serverIpAddress.mFields.m8[addrSize - 2] = static_cast<uint8_t>((aRloc16 & 0xff00) >> 8u);
```
Wouldn't something like `serverIpAddress.mFields.m16[7] = aRloc16` be better, alternatively if endianess would be different, don't we have sth like NTOHS or HostSwap16() function in CHIP..?

LuDuda commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +998,237 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+    return MapOpenThreadError(OT_ERROR_NONE);
+}
+
+#endif // CHIP_DEVICE_CONFIG_ENABLE_THREAD_SRP_CLIENT && OPENTHREAD_API_VERSION >= 72
```
```suggestion
#endif // CHIP_DEVICE_CONFIG_ENABLE_THREAD_SRP_CLIENT
```

LuDuda commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.h:
```diff
@@ -26,6 +26,8 @@
 
 #include <openthread/instance.h>
+#include <openthread/netdata.h>
+#include <openthread/srp_client.h>
```
Same comment as above.

LuDuda commented on src/platform/nrfconnect/CHIPDevicePlatformConfig.h:
```diff
@@ -75,3 +75,8 @@
 #endif // CONFIG_CHIP_ENABLE_RENDEZVOUS_BYPASS
+
+#ifdef CONFIG_CHIP_ENABLE_SERVICE_REGISTRATION
+#define CHIP_DEVICE_CONFIG_ENABLE_MDNS 1
```
Hmm.. does `CHIP_DEVICE_CONFIG_ENABLE_MDNS` needs to be defined? Maybe it should be renamed to sth like `DNSSD` and `MDNS` where applicable - but I haven't looked deeply in the code.

kkasperczyk-no replied:
Unfortunately yes, we decided with Andrei for now to make from SRP something like service discovery implementation for MDNS. The goal is to reuse as much MDNS code and structures as possible and keep it consistent. I guess it is temporary solution to introduce SRP in CHIP, as soon as possible, but I talked with Damian that we should plan to refactor it in next sprints and create more generic module.

LuDuda replied:
Okay.

LuDuda reviewed (changes requested):
Good work, Kamil.

Similarily to Damian, i have few nits/preferences and few more serious comments. I think you should take a look on srp_client.h API again, since many of the parameters are expected to be persistent, while you allocate few things on stack.

kkasperczyk-no commented on src/include/platform/CHIPDeviceConfig.h:
```diff
@@ -522,6 +522,26 @@
+ * Enable support to DNS-SD SRP client usage for service advertising and discovery in CHIP.
+ */
+#ifndef CHIP_DEVICE_CONFIG_ENABLE_THREAD_SRP_CLIENT
+#define CHIP_DEVICE_CONFIG_ENABLE_THREAD_SRP_CLIENT 0
```
Actually I added THREAD word on @Damian-Nordic request, as he suggested that SRP probably will be used only for Thread, what is kind of true.

Damian-Nordic replied:
Yes, "THREAD" is my fault :) I wanted it as the define only controls `ThreadStackManager` so I thought it would be more informative that way.

kkasperczyk-no commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +998,237 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+    while (otNetDataGetNextService(mOTInst, &iterator, &aServiceConfig) == OT_ERROR_NONE)
+    {
+        if (aServiceConfig.mEnterpriseNumber == kThreadServiceEnterpriseNumber &&
+            aServiceConfig.mServiceDataLength == sizeof(kSrpServiceId) && aServiceConfig.mServiceData[0] == kSrpServiceId)
```
After moving it to the new line, clang-format is moving it back there, so I guess we should leave it as is.

kkasperczyk-no commented on config/nrfconnect/chip-module/Kconfig:
```diff
@@ -94,6 +94,14 @@ config CHIP_ENABLE_RENDEZVOUS_BYPASS
 
+config CHIP_ENABLE_SERVICE_REGISTRATION
+	bool "Enable support for service registration"
+	default n
```
How about use cases like light bulb & light switch samples? Will we align them to use SRP? I assumed that in some test cases we won't use SRP, but of course I can reverse it, make option enabled by default and we will disable it when needed. I just usually prefer making something disabled by default and enabling when necessary :)

LuDuda replied:
Yeah, it may be just a matter of preference. But i would set it by default due to the fact it is the mandatory feature in CHIP.

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +1002,280 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+            {
+                next = service->mNext;
+                // Clear also names associated with the SrpClient::Service object
+                memset(const_cast<char *>(service->mInstanceName), 0, strlen(service->mInstanceName) + 1);
```
How about you simply assign `service->mInstanceName = nullptr` or `memset(service, 0, sizeof(otSrpClientService));` and in the function looking for a free slot you compare `mService.mInstanceName`? Should be a little bit more efficient and will the outcome should be the same.

kkasperczyk-no replied:
I thought about it, but I decided on current solution, as it's more transparent when reading code. It we do as you suggest, we will meet a situation when for some time Service has assigned mName and mInstanceName, but actually otSrpClientService does not exist. It will not lead to any faults for now, but I imagine that someone will try to edit code in the future and will be pretty much confused that data are not consistent.

LuDuda replied:
Hmm I'm also not convinced.. but If we need to clear this, maybe just `service->mInstanceName[0] = '\0'`, same with `mName`?

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +1002,280 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+        }
+        else
+        {
+            VerifyOrExit(strcmp(service.mInstanceName, aInstanceName) != 0, error = MapOpenThreadError(OT_ERROR_DUPLICATED));
```
Would it be difficult to update the service in such a case? Also, probably both aName and aInstanceName should match to get a duplicate.

kkasperczyk-no replied:
For now it doesn't seem to work, I will need to ask SRP authors for it. Calling otSrpClientAddService() for the same service with different values doesn't result in changing those values. In the OT CLI example modifying services works, but it is done by adding the same service multiple times with different values (e.g. port number). For example modifying the added service 2 times results in having the same 3 services in an array and only one on a server. When array has max size of 3 elements it means, that we have full array, while server still sees only 1 service. IMO it doesn't seem to be a good solution.

Damian-Nordic replied:
Ok, sure, we can look at that later!

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +1002,280 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+    // Verify is there an empty place for new service.
+    VerifyOrExit(srpService, error = MapOpenThreadError(OT_ERROR_NO_BUFS));
+
+    otSrpClientSetLeaseInterval(mOTInst, aLeaseInterval);
```
Isn't there a way to configure the lease times per service? Should we check if all registered services don't have conflicting demands?

kkasperczyk-no replied:
AFAIK currently there is no way to do that. If you set lease time it applies to all services added after that operation, but it doesn't change anything for the ones added previously. That's why I decided to call otSrpClientSetLeaseInterval on adding each service, what allows user to call AddSrpService with different lease time value.

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +1002,280 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+template <class ImplClass>
+uint16_t GenericThreadStackManagerImpl_OpenThread<ImplClass>::ConvertSrpServerDataToPort(uint8_t * aServerData)
+{
+    static_assert(OT_SERVER_DATA_MAX_SIZE >= 2, "");
```
Now that you verify the actual length of the data, it can probably be skipped.

Damian-Nordic reviewed (approved):
Very nice!

kkasperczyk-no commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.h:
```diff
@@ -99,6 +113,36 @@ class GenericThreadStackManagerImpl_OpenThread
+
+        struct Service
+        {
+            bool isEmpty;
```
Oh right, removed it.

kkasperczyk-no commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -992,6 +993,233 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+    VerifyOrExit(aHostName, error = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(!otSrpClientIsRunning(mOTInst), error = OT_ERROR_ALREADY);
+
+    memcpy(&clientIpAddress, otThreadGetMeshLocalEid(mOTInst), sizeof(clientIpAddress));
```
I replaced GetSlaacIPv6Address with GetExternalIPv6Address considering addresses assigned by SLAAC, DHCPv6 or MANUAL. Name of GetExternalIPv6Address should be discussed and maybe aligned in the future, as it is not proper nomenclature, but it also can't be name GetGlobalIPv6Address, as addresses with prefix `fd` don't meet with GUA requirements according to the: https://openthread.io/guides/thread-primer/ipv6-addressing
> Global Unicast Address (GUA) - Always has a prefix of 2000::/3
>

LuDuda commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -792,14 +797,27 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_GetPrimary80215
+    for (const otNetifAddress * otAddress = otAddresses; otAddress != nullptr; otAddress = otAddress->mNext)
     {
-        if (otAddr->mValid && otAddr->mAddressOrigin == OT_ADDRESS_ORIGIN_SLAAC)
+        if (otAddress->mValid && otAddress->mAddressOrigin)
```
Hmm why to do `&& otAddress->mAddressOrigin` we anyway check `OT_ADDRESS_ORIGIN_THREAD == 0` in switch case below.

LuDuda reviewed (approved):
Looks good! Thanks :+1:

kkasperczyk-no commented:
@LuDuda @Damian-Nordic please re-review that one. Start/Stop SRP Client and few helper methods used for conversion purposes were removed. Added enabling SRP client autostart on init and SetSrpHost method, as user needs to pass host name to it, even though client is started automatically. I implemented also OnSrpClientStateChange callback informing about following starts/stops of SRP client.

Damian-Nordic commented on src/include/platform/CHIPDeviceConfig.h:
```diff
@@ -532,6 +532,26 @@
+ *
+ * Enable support to use MDNS for service advertising and discovery in CHIP.
+ */
+#ifndef CHIP_DEVICE_CONFIG_ENABLE_MDNS
```
After second thought, maybe it would be better if we set GN's chip_enable_mdns based on Kconfig, and not this define. The reason is that `chip_enable_mdns` not only affects `CHIP_DEVICE_CONFIG_ENABLE_MDNS`, but can also be used to control which files are built. What do you think?

Damian-Nordic replied:
Or we can always fixed when it's needed... Nevermind then.

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -1002,6 +1026,211 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+            {
+                next = service->mNext;
+                // Clear also names associated with the SrpClient::Service object
+                service->mInstanceName[0] = '\0';
```
The code doesn't do what you probably intended it to do. It only clears the pointer which you anyway clear in the following `memset`.

kkasperczyk-no replied:
I'm not sure I understand. `service` is `otSrpClientService` type and it has a field mInstanceName, which is a pointer set to point at `Service mInstanceName` array. So when we set `service->mInstanceName[0] = '\0\'`, it actually sets the first element of `Service mInstanceName` array. It is not the same as setting `service->mInstanceName` to nullptr, as it will just remove address from pointer and will not clear `Service mInstanceName` array. Then when calling `if (strcmp(service.mInstanceName, "") == 0)` we got the first element equal to `'\0'` and condition is true.

Damian-Nordic replied:
Right, I must be tired :) It's however weird that the compiler allows you to modify a constant. Or in newer OT it's no longer `const char*`? https://openthread.io/reference/struct/ot-srp-client-service?authuser=1#minstancename

kkasperczyk-no replied:
It is constant, the thing is that we do not modify it :) Pointer is just an address of object, so we do not change address, but only object under that address. So from the pointer point of view, we do not modify it.

Damian-Nordic replied:
But `const char*` is a pointer to constant. The code:
```
void f(const char* a)
{
    a[0] = '\0';
}
```
won't build. So why your code does? :)

kkasperczyk-no replied:
Oh now I see, I guess we both were looking on arguments, but here I did const_cast https://github.com/kkasperczyk-no/connectedhomeip/pull/21/files#diff-837473660e25fab2523b543759412eb6ccd77f7c790b41126811387fa7916b20R1047

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -1002,6 +1026,211 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+    switch (aError)
+    {
+    case OT_ERROR_NONE: {
+        ChipLogProgress(DeviceLayer, "OnSrpClientNotification: Last requested operation completed successfully");
```
Doesn't it flood the logs with unnecessary information? If so, consider `ChipLogDetail`.

kkasperczyk-no replied:
Not exactly, as probably it will be displayed only once, per each service. And in most cases we will use only one service for SRP.

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -1002,6 +1026,211 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+    if (aServerSockAddr)
+    {
+        ChipLogProgress(DeviceLayer,
+                        "OnSrpClientStateChange: SRP Client was started, as detected server addressed: %x:%x:%x:%x:%x:%x:%x:%x",
```
Does such long message print correctly? I think Zephyr has some limits so it's worth checking or shorten the message (e.g. by removing the OnSrpClientStateChange: prefix).

kkasperczyk-no replied:
Yeah it prints correctly, however I shortened it a little just in case.

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:
```diff
@@ -1002,6 +1026,211 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+    // Check if service with desired instance name already exists and try to find empty slot in array for new service
+    for (typename SrpClient::Service & service : mSrpClient.mServices)
+    {
+        if (strcmp(service.mInstanceName, "") == 0)
```
Fix the clearing issue mentioned above or check `service.mService.mInstanceName == nullptr` here.

Damian-Nordic commented on src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.h:
```diff
@@ -100,6 +113,37 @@ class GenericThreadStackManagerImpl_OpenThread
+
+    struct SrpClient
+    {
+        static constexpr uint16_t kThreadServiceEnterpriseNumber = 44970;
```
No longer needed?
